C# analysis failing hard when source generators are used

Reported on the community forum: https://community.sonarsource.com/t/c-analysis-failing-in-azure-pipelines-due-to-java-lang-illegalargumentexception-uri-is-not-absolute/38995 . Project reproducer, logs and all information is present there.


> The .CommandGenerator.cs files are indeed generated during build time. This is triggered by using the [GenerateCommand] attribute in WpfDemo/DemoViewModel.cs.

The SARIF report parsing is failing hard:

```
2021-02-23T21:59:23.4779322Z WARN: File 'Generators\Generators.CommandGenerator\GenerateCommandAttribute.cs' referenced by the protobuf 'MetricsInfo' does not exist in the analysis context
2021-02-23T21:59:23.5134921Z WARN: File 'Generators\Generators.CommandGenerator\DemoViewModel.CommandGenerator.cs' referenced by the protobuf 'MetricsInfo' does not exist in the analysis context
2021-02-23T21:59:23.5136692Z WARN: File 'Generators\Generators.CommandGenerator\GenerateCommandAttribute.cs' referenced by the protobuf 'TokenTypeInfo' does not exist in the analysis context
2021-02-23T21:59:23.5138016Z WARN: File 'Generators\Generators.CommandGenerator\DemoViewModel.CommandGenerator.cs' referenced by the protobuf 'TokenTypeInfo' does not exist in the analysis context
2021-02-23T21:59:23.5139426Z WARN: File 'Generators\Generators.CommandGenerator\GenerateCommandAttribute.cs' referenced by the protobuf 'SymbolReferenceInfo' does not exist in the analysis context
2021-02-23T21:59:23.5142730Z WARN: File 'Generators\Generators.CommandGenerator\DemoViewModel.CommandGenerator.cs' referenced by the protobuf 'SymbolReferenceInfo' does not exist in the analysis context
2021-02-23T21:59:23.5144461Z WARN: File 'Generators\Generators.CommandGenerator\GenerateCommandAttribute.cs' referenced by the protobuf 'CopyPasteTokenInfo' does not exist in the analysis context
2021-02-23T21:59:23.5145858Z WARN: File 'Generators\Generators.CommandGenerator\DemoViewModel.CommandGenerator.cs' referenced by the protobuf 'CopyPasteTokenInfo' does not exist in the analysis context
2021-02-23T21:59:23.5240639Z INFO: Importing 2 Roslyn reports
2021-02-23T21:59:23.8892188Z INFO: ------------------------------------------------------------------------
2021-02-23T21:59:23.9082600Z INFO: EXECUTION FAILURE
2021-02-23T21:59:23.9099757Z INFO: ------------------------------------------------------------------------
2021-02-23T21:59:23.9103004Z INFO: Total time: 43.754s
2021-02-23T21:59:23.9116543Z INFO: Final Memory: 27M/90M
2021-02-23T21:59:23.9118045Z INFO: ------------------------------------------------------------------------
2021-02-23T21:59:23.9119856Z ##[error]ERROR: Error during SonarScanner execution
2021-02-23T21:59:23.9121605Z ERROR: Error during SonarScanner execution
2021-02-23T21:59:23.9123175Z ##[error]java.lang.IllegalArgumentException: URI is not absolute
2021-02-23T21:59:23.9128841Z java.lang.IllegalArgumentException: URI is not absolute
2021-02-23T21:59:23.9130496Z ##[error]at java.base/java.io.File.<init>(File.java:418)
2021-02-23T21:59:23.9134877Z 	at java.base/java.io.File.<init>(File.java:418)
2021-02-23T21:59:23.9136986Z ##[error]at org.sonarsource.dotnet.shared.sarif.SarifParser10.uriToAbsolutePath(SarifParser10.java:225)
2021-02-23T21:59:23.9139038Z 	at org.sonarsource.dotnet.shared.sarif.SarifParser10.uriToAbsolutePath(SarifParser10.java:225)
2021-02-23T21:59:23.9140788Z ##[error]at org.sonarsource.dotnet.shared.sarif.SarifParser10.handleLocation(SarifParser10.java:175)
2021-02-23T21:59:23.9142659Z 	at org.sonarsource.dotnet.shared.sarif.SarifParser10.handleLocation(SarifParser10.java:175)
2021-02-23T21:59:23.9144410Z ##[error]at org.sonarsource.dotnet.shared.sarif.SarifParser10.handleResultFileElement(SarifParser10.java:146)
2021-02-23T21:59:23.9146191Z 	at org.sonarsource.dotnet.shared.sarif.SarifParser10.handleResultFileElement(SarifParser10.java:146)
2021-02-23T21:59:23.9147940Z ##[error]at org.sonarsource.dotnet.shared.sarif.SarifParser10.handleLocationsElement(SarifParser10.java:138)
2021-02-23T21:59:23.9149709Z 	at org.sonarsource.dotnet.shared.sarif.SarifParser10.handleLocationsElement(SarifParser10.java:138)
2021-02-23T21:59:23.9151397Z ##[error]at org.sonarsource.dotnet.shared.sarif.SarifParser10.handleIssue(SarifParser10.java:107)
2021-02-23T21:59:23.9153212Z 	at org.sonarsource.dotnet.shared.sarif.SarifParser10.handleIssue(SarifParser10.java:107)
2021-02-23T21:59:23.9155243Z ##[error]at org.sonarsource.dotnet.shared.sarif.SarifParser10.handleIssues(SarifParser10.java:95)
2021-02-23T21:59:23.9157486Z 	at org.sonarsource.dotnet.shared.sarif.SarifParser10.handleIssues(SarifParser10.java:95)
2021-02-23T21:59:23.9159270Z ##[error]at org.sonarsource.dotnet.shared.sarif.SarifParser10.accept(SarifParser10.java:65)
2021-02-23T21:59:23.9161073Z 	at org.sonarsource.dotnet.shared.sarif.SarifParser10.accept(SarifParser10.java:65)
2021-02-23T21:59:23.9162930Z ##[error]at org.sonarsource.dotnet.shared.sarif.SarifParser10.accept(SarifParser10.java:39)
2021-02-23T21:59:23.9164756Z 	at org.sonarsource.dotnet.shared.sarif.SarifParser10.accept(SarifParser10.java:39)
2021-02-23T21:59:23.9166545Z ##[error]at org.sonarsource.dotnet.shared.plugins.RoslynDataImporter.importRoslynReports(RoslynDataImporter.java:54)
2021-02-23T21:59:23.9168423Z 	at org.sonarsource.dotnet.shared.plugins.RoslynDataImporter.importRoslynReports(RoslynDataImporter.java:54)
2021-02-23T21:59:23.9170358Z ##[error]at org.sonarsource.dotnet.shared.plugins.DotNetSensor.executeInternal(DotNetSensor.java:111)
2021-02-23T21:59:23.9173852Z 	at org.sonarsource.dotnet.shared.plugins.DotNetSensor.executeInternal(DotNetSensor.java:111)
2021-02-23T21:59:23.9175845Z ##[error]at org.sonarsource.dotnet.shared.plugins.DotNetSensor.execute(DotNetSensor.java:71)
2021-02-23T21:59:23.9177777Z 	at org.sonarsource.dotnet.shared.plugins.DotNetSensor.execute(DotNetSensor.java:71)
```

--- a/sonar-dotnet-shared-library/src/main/java/org/sonarsource/dotnet/shared/sarif/SarifParser10.java
+++ b/sonar-dotnet-shared-library/src/main/java/org/sonarsource/dotnet/shared/sarif/SarifParser10.java
@@ -147,8 +147,8 @@ class SarifParser10 implements SarifParser {
 
     if (primaryLocation == null) {
       String uri = resultFileObj.get("uri").getAsString();
-      String absolutePath = toRealPath.apply(uriToAbsolutePath(uri));
-      callback.onFileIssue(ruleId, level, absolutePath, message);
+      String path = toRealPath.apply(uriToPath(uri));
+      callback.onFileIssue(ruleId, level, path, message);
     } else {
       Collection<Location> secondaryLocations = new ArrayList<>();
       for (JsonElement relatedLocationEl : relatedLocations) {
@@ -172,7 +172,7 @@ class SarifParser10 implements SarifParser {
   @CheckForNull
   private Location handleLocation(JsonObject locationObj, @Nullable String message) {
     String uri = locationObj.get("uri").getAsString();
-    String absolutePath = toRealPath.apply(uriToAbsolutePath(uri));
+    String path = toRealPath.apply(uriToPath(uri));
     JsonObject region = locationObj.get("region").getAsJsonObject();
     int startLine = region.get("startLine").getAsInt();
 
@@ -182,7 +182,7 @@ class SarifParser10 implements SarifParser {
 
     JsonElement lengthOrNull = region.get("length");
     if (lengthOrNull != null) {
-      return new Location(absolutePath, message, startLine, startLineOffset, startLine, startLineOffset + lengthOrNull.getAsInt());
+      return new Location(path, message, startLine, startLineOffset, startLine, startLineOffset + lengthOrNull.getAsInt());
     }
 
     JsonElement endLineOrNull = region.get("endLine");
@@ -204,7 +204,7 @@ class SarifParser10 implements SarifParser {
       return null;
     }
 
-    return new Location(absolutePath, message, startLine, startLineOffset, endLine, endLineOffset);
+    return new Location(path, message, startLine, startLineOffset, endLine, endLineOffset);
   }
 
   private static boolean isSuppressed(JsonObject resultObj) {
@@ -220,9 +220,14 @@ class SarifParser10 implements SarifParser {
     return false;
   }
 
-  private static String uriToAbsolutePath(String uri) {
+  private static String uriToPath(String uri) {
     String uriEscaped = uri.replace("[", "%5B").replace("]", "%5D");
-    return new File(URI.create(uriEscaped)).getAbsolutePath();
+    String uriPath = URI.create(uriEscaped).getPath();
+    File file = new File(uriPath);
+
+    return file.isAbsolute()
+      ? file.getAbsolutePath()
+      : file.getPath();
   }
 
 }

--- a/sonar-dotnet-shared-library/src/test/java/org/sonarsource/dotnet/shared/sarif/SarifParser10Test.java
+++ b/sonar-dotnet-shared-library/src/test/java/org/sonarsource/dotnet/shared/sarif/SarifParser10Test.java
@@ -226,4 +226,17 @@ public class SarifParser10Test {
     verifyNoMoreInteractions(callback);
   }
 
+  @Test
+  public void sarif_version_1_0_relative_paths() throws IOException {
+    SarifParserCallback callback = mock(SarifParserCallback.class);
+    new SarifParser10(null, getRoot("v1_0_relative_paths.json"), String::toString).accept(callback);
+    InOrder inOrder = inOrder(callback);
+
+    Location s1144PrimaryLocation = new Location("SourceGeneratorPOC.Generators\\SourceGeneratorPOC.SourceGenerator\\Greetings.cs", "Remove the unused private method 'UnusedMethod'.", 7, 8, 7, 46);
+    inOrder.verify(callback).onIssue("S1144", "warning", s1144PrimaryLocation, new ArrayList<>());
+
+    Location s1186PrimaryLocation = new Location("SourceGeneratorPOC.Generators\\SourceGeneratorPOC.SourceGenerator\\Greetings.cs", "Add a nested comment explaining why this method is empty, throw a 'NotSupportedException' or complete the implementation.", 7, 28, 7, 40);
+    inOrder.verify(callback).onIssue("S1186", "warning", s1186PrimaryLocation, new ArrayList<>());
+  }
+
 }

--- a/sonar-dotnet-shared-library/src/test/java/org/sonarsource/dotnet/shared/sarif/SarifParser10Test.java
+++ b/sonar-dotnet-shared-library/src/test/java/org/sonarsource/dotnet/shared/sarif/SarifParser10Test.java
@@ -239,4 +239,14 @@ public class SarifParser10Test {
     inOrder.verify(callback).onIssue("S1186", "warning", s1186PrimaryLocation, new ArrayList<>());
   }
 
+  @Test
+  public void sarif_version_1_0_region_with_length() throws IOException {
+    SarifParserCallback callback = mock(SarifParserCallback.class);
+    new SarifParser10(null, getRoot("v1_0_region_with_length.json"), String::toString).accept(callback);
+    InOrder inOrder = inOrder(callback);
+
+    Location primaryLocation = new Location("SourceGeneratorPOC.Generators\\SourceGeneratorPOC.SourceGenerator\\Greetings.cs", "Remove the unused private method 'UnusedMethod'.", 7, 8, 7, 46);
+    inOrder.verify(callback).onIssue("S1144", "warning", primaryLocation, new ArrayList<>());
+  }
+
 }

--- a/sonar-dotnet-shared-library/src/test/resources/SarifParserTest/v1_0_region_with_length.json
+++ b/sonar-dotnet-shared-library/src/test/resources/SarifParserTest/v1_0_region_with_length.json
@@ -1,0 +1,58 @@
+{
+  "$schema": "http://json.schemastore.org/sarif-1.0.0",
+  "version": "1.0.0",
+  "runs": [
+    {
+      "tool": {
+        "name": "Microsoft (R) Visual C# Compiler",
+        "version": "3.9.0.0",
+        "fileVersion": "3.9.0-5.21120.8 (accdcb77)",
+        "semanticVersion": "3.9.0",
+        "language": "en-US"
+      },
+      "results": [
+        {
+          "ruleId": "S1144",
+          "level": "warning",
+          "message": "Remove the unused private method 'UnusedMethod'.",
+          "locations": [
+            {
+              "resultFile": {
+                "uri": "SourceGeneratorPOC.Generators/SourceGeneratorPOC.SourceGenerator/Greetings.cs",
+                "region": {
+                  "startLine": 7,
+                  "startColumn": 9,
+                  "endLine": 7,
+                  "length": 38
+                }
+              }
+            }
+          ],
+          "properties": {
+            "warningLevel": 1
+          }
+        }
+      ],
+      "rules": {
+        "S1144": {
+          "id": "S1144",
+          "shortDescription": "Unused private types or members should be removed",
+          "fullDescription": "private or internal types or private members that are never executed or referenced are dead code: unnecessary, inoperative code that should be removed. Cleaning out dead code decreases the size of the maintained codebase, making it easier to understand the program and preventing bugs from being introduced.",
+          "defaultLevel": "note",
+          "helpUri": "https://rules.sonarsource.com/csharp/RSPEC-1144",
+          "properties": {
+            "category": "Major Code Smell",
+            "isEnabledByDefault": true,
+            "tags": [
+              "C#",
+              "SonarWay",
+              "MainSourceScope",
+              "TestSourceScope",
+              "Unnecessary"
+            ]
+          }
+        }
+      }
+    }
+  ]
+}

--- a/sonar-dotnet-shared-library/src/test/resources/SarifParserTest/v1_0_relative_paths.json
+++ b/sonar-dotnet-shared-library/src/test/resources/SarifParserTest/v1_0_relative_paths.json
@@ -1,0 +1,96 @@
+{
+  "$schema": "http://json.schemastore.org/sarif-1.0.0",
+  "version": "1.0.0",
+  "runs": [
+    {
+      "tool": {
+        "name": "Microsoft (R) Visual C# Compiler",
+        "version": "3.9.0.0",
+        "fileVersion": "3.9.0-5.21120.8 (accdcb77)",
+        "semanticVersion": "3.9.0",
+        "language": "en-US"
+      },
+      "results": [
+        {
+          "ruleId": "S1144",
+          "level": "warning",
+          "message": "Remove the unused private method 'UnusedMethod'.",
+          "locations": [
+            {
+              "resultFile": {
+                "uri": "SourceGeneratorPOC.Generators/SourceGeneratorPOC.SourceGenerator/Greetings.cs",
+                "region": {
+                  "startLine": 7,
+                  "startColumn": 9,
+                  "endLine": 7,
+                  "endColumn": 47
+                }
+              }
+            }
+          ],
+          "properties": {
+            "warningLevel": 1
+          }
+        },
+        {
+          "ruleId": "S1186",
+          "level": "warning",
+          "message": "Add a nested comment explaining why this method is empty, throw a 'NotSupportedException' or complete the implementation.",
+          "locations": [
+            {
+              "resultFile": {
+                "uri": "SourceGeneratorPOC.Generators/SourceGeneratorPOC.SourceGenerator/Greetings.cs",
+                "region": {
+                  "startLine": 7,
+                  "startColumn": 29,
+                  "endLine": 7,
+                  "endColumn": 41
+                }
+              }
+            }
+          ],
+          "properties": {
+            "warningLevel": 1
+          }
+        }
+      ],
+      "rules": {
+        "S1144": {
+          "id": "S1144",
+          "shortDescription": "Unused private types or members should be removed",
+          "fullDescription": "private or internal types or private members that are never executed or referenced are dead code: unnecessary, inoperative code that should be removed. Cleaning out dead code decreases the size of the maintained codebase, making it easier to understand the program and preventing bugs from being introduced.",
+          "defaultLevel": "note",
+          "helpUri": "https://rules.sonarsource.com/csharp/RSPEC-1144",
+          "properties": {
+            "category": "Major Code Smell",
+            "isEnabledByDefault": true,
+            "tags": [
+              "C#",
+              "SonarWay",
+              "MainSourceScope",
+              "TestSourceScope",
+              "Unnecessary"
+            ]
+          }
+        },
+        "S1186": {
+          "id": "S1186",
+          "shortDescription": "Methods should not be empty",
+          "fullDescription": "There are several reasons for a method not to have a method body:",
+          "defaultLevel": "warning",
+          "helpUri": "https://rules.sonarsource.com/csharp/RSPEC-1186",
+          "properties": {
+            "category": "Critical Code Smell",
+            "isEnabledByDefault": true,
+            "tags": [
+              "C#",
+              "SonarWay",
+              "MainSourceScope",
+              "TestSourceScope"
+            ]
+          }
+        }
+      }
+    }
+  ]
+}

